### PR TITLE
wip TMEDIA-671 - caption component

### PR DESCRIPTION
**Pull request by JackHowa**
# Question

- Is this component properties the naming scheme and class names that we're imagining? See scss file and index js file for caption. I looked in the poc to look for these but couldn't find any. -> yes it's ok naming
- 
```
Includes three component properties: Title, Caption, and Credit

Caption - Title

Caption - Caption

Caption - Credit
```

# Todo 

- [ ] Add format credit util function 
- [ ] Fill out description and usage
- [ ] Add potential nodes as props for taking in title, etc (eg, title={`<div>Here's a <b>title</b></div>}`)
- [ ] Include more tests for no presence of title, other props
- [ ] Add more thorough title and caption descriptions

### CODE REVIEW

File: src/components/caption/_index.scss [hunk 1]
```diff
@@ -0,0 +1,17 @@
+@use "../../scss";
+
+.c-caption {
+	@include scss.component-properties("caption");
+
+	.c-caption__title {
```
Comment (JackHowa) [inline src/components/caption/_index.scss:6]: classname with bem element ok?

File: src/components/caption/_index.scss [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+		@include scss.component-properties("caption-title");
```
Comment (JackHowa) [inline src/components/caption/_index.scss:7]: is this naming scheme ok `caption-title` for component properties?

File: src/components/caption/_index.scss (hunk 1 above)
Comment (matthewroach) [reply src/components/caption/_index.scss:6]: I don't see a reason not too

File: src/components/caption/_index.scss (hunk 2 above)
Comment (matthewroach) [reply src/components/caption/_index.scss:7]: Yeah this works and it's something that has been done in other components

File: src/components/caption/index.jsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import PropTypes from "prop-types";
+
+const COMPONENT_CLASS_NAME = "c-caption";
+
+const Caption = ({ caption, className, credit, title }) =>
+	!!(title || caption || credit) && (
+		<figcaption
+			className={className ? `${COMPONENT_CLASS_NAME} ${className}` : `${COMPONENT_CLASS_NAME}`}
+		>
+			{title && <span className="c-caption__title">{title} </span>}
```
Comment (matthewroach) [inline src/components/caption/index.jsx]: We want to make sure we are using ternary vs logical on rendering `{title ? <span></span> : null}`

File: src/components/caption/index.jsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+			{caption && <span className="c-caption__caption">{caption} </span>}
```
Comment (matthewroach) [inline src/components/caption/index.jsx]: Also note the requirement that these items can have HTML within them and will need to render that HTML

File: src/components/caption/index.test.jsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+		render(<Caption title="title" credit="credit" caption="caption" />);
+
+		expect(screen.queryByText("title")).not.toBeNull();
+		expect(screen.queryByText("credit")).not.toBeNull();
+		expect(screen.queryByText("caption")).not.toBeNull();
+	});
+	it("should render string child with additional classes", () => {
+		const { container } = render(<Caption title="Hello World" className="test-class" />);
+		// not quite best practice with rtl's methodology of not testing for classes
+		// couldn't find a way to test for className without using its role or text
```
Comment (nschubach) [inline src/components/caption/index.test.jsx:16]: testing-library (basis of RTL) has `.toHaveClass('')` ... I think people get hung up on the "not testing implementation" to include classes, but classes are used to define appearance which is what the library claims is the best way to test.

https://github.com/testing-library/jest-dom#tohaveclass

File: src/components/caption/_index.scss (hunk 2 above)
Comment (JackHowa) [reply src/components/caption/_index.scss:7]: oh ok. I didn't realize this was synonymous with other :hover state. sounds good

File: src/components/caption/index.jsx (hunk 4 above)
Comment (JackHowa) [reply src/components/caption/index.jsx]: Yep, thanks, that's included in my todos

File: src/components/caption/index.test.jsx (hunk 5 above)
Comment (JackHowa) [reply src/components/caption/index.test.jsx:16]: gotcha gotcha -- thanks. saw this in your generator pr as well

File: src/components/caption/index.jsx (hunk 3 above)
Comment (JackHowa) [reply src/components/caption/index.jsx]: updated

File: package.json [hunk 6]
```diff
@@ -56,6 +56,7 @@
     "babel-jest": "^27.4.6",
     "chromatic": "^6.4.3",
     "css-loader": "^5.2.7",
+    "dompurify": "^2.3.6",
```
Comment (matthewroach) [inline package.json:59]: We are adding a sanitising library? We had discussed not doing this within these components in other tickets. Was there a change?

Comment (brentmiller1973) [reply package.json:59]: @matthewroach You're suggesting then that sanitising would then take place on the block level?

Comment (matthewroach) [reply package.json:59]: Currently we don't do any sanitising, and that was one of the reasons we've yet to decided where it would be done

Comment (JackHowa) [reply package.json:59]: gotcha no problem

**Pull request closed without merging**